### PR TITLE
Update com.google.errorprone:* to 2.31.0

### PR DESCRIPTION
**Pull request by malliaridis**
# Description

This PR updates Google error-prone to 2.31.0.

# Solution

In order to avoid blocking https://github.com/apache/solr/pull/2706 I have updated Google error-prone and added all new checks introduced between 2.23.0 and 2.31.0 without enabling them to `gradle/validation/error-prone.gradle`. Enabling the new rules should be handled in a separate PR as this may require further evaluation and compatibility check and would further delay https://github.com/apache/solr/pull/2706.

The selected version 2.31.0 is the last version of error-prone that fully supports JDK 11, so this PR should be backwards compatible. Note that some new rules are not applicable to JDK 11 and should carefully be picked upon activation.

Some of the existing rules seem to have had some changes that caused existing checks to fail. For the affected code, I decided individually wether to suppress the warning or fix the code.

See commit details for additional information about individual changes.

# Tests

Some tests were affected by this update and were updated. The updates include  assertions  that were checking primitive data types for nullability and were therefore removed.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://github.com/apache/solr/blob/main/CONTRIBUTING.md) and my code conforms to the standards described there to the best of my ability.
- [ ] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended, not available for branches on forks living under an organisation)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

File: gradle/validation/error-prone.gradle [hunk 1]
```diff
@@ -236,7 +240,7 @@ allprojects { prj ->
             // '-Xep:ThrowIfUncheckedKnownChecked:OFF', // we don't use this annotation
             '-Xep:ThrowNull:ERROR',
             '-Xep:TreeToString:ERROR',
-            // '-Xep:TruthSelfEquals:OFF', // we don't use truth
+            // '-Xep:SelfAssertion:OFF', // we don't use truth
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:243]: SelfAssertion is listed twice now? Line 226

**Review by risdenk — APPROVED** [review]

File: gradle/validation/error-prone.gradle [hunk 2]
```diff
@@ -179,6 +179,7 @@ allprojects { prj ->
             '-Xep:MathRoundIntLong:ERROR',
             // '-Xep:MislabeledAndroidString:OFF', // we don't use android
             '-Xep:MisplacedScopeAnnotations:ERROR',
+            // '-Xep:MissingRuntimeRetention:ERROR', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:182]: We don't use guice so this can be ignored.

File: gradle/validation/error-prone.gradle [hunk 3]
```diff
@@ -218,12 +219,15 @@ allprojects { prj ->
             '-Xep:RandomCast:ERROR',
             '-Xep:RandomModInteger:ERROR',
             // '-Xep:RectIntersectReturnValueIgnored:OFF', // we don't use android
+            // '-Xep:RedundantSetterCall:ERROR', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:222]: We don't use AutoValue

File: gradle/validation/error-prone.gradle [hunk 4]
```diff
@@ -218,12 +219,15 @@ allprojects { prj ->
...
             // '-Xep:RequiredModifiers:OFF', // we don't use this annotation
             // '-Xep:RestrictedApiChecker:OFF', // we don't use this annotation
             // '-Xep:ReturnValueIgnored:OFF', // todo there are problems that should be fixed
+            // '-Xep:SelfAssertion:ERROR', // todo check if useful or comment why not
             '-Xep:SelfAssignment:ERROR',
             '-Xep:SelfComparison:ERROR',
             '-Xep:SelfEquals:ERROR',
+            // '-Xep:SetUnrecognized:ERROR', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:230]: We don't use proto so can be ignored

File: gradle/validation/error-prone.gradle [hunk 5]
```diff
@@ -218,12 +219,15 @@ allprojects { prj ->
             '-Xep:RandomCast:ERROR',
             '-Xep:RandomModInteger:ERROR',
             // '-Xep:RectIntersectReturnValueIgnored:OFF', // we don't use android
+            // '-Xep:RedundantSetterCall:ERROR', // todo check if useful or comment why not
             // '-Xep:RequiredModifiers:OFF', // we don't use this annotation
             // '-Xep:RestrictedApiChecker:OFF', // we don't use this annotation
             // '-Xep:ReturnValueIgnored:OFF', // todo there are problems that should be fixed
+            // '-Xep:SelfAssertion:ERROR', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:226]: We don't use truth so can be ignored

File: gradle/validation/error-prone.gradle [hunk 6]
```diff
@@ -265,6 +268,7 @@ allprojects { prj ->
             '-Xep:AssertionFailureIgnored:WARN',
             '-Xep:AssistedInjectAndInjectOnSameConstructor:WARN',
             '-Xep:AttemptedNegativeZero:WARN',
+            // '-Xep:AutoValueBoxedValues:WARN', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:271]: This is probably useful

File: gradle/validation/error-prone.gradle [hunk 7]
```diff
@@ -285,6 +289,7 @@ allprojects { prj ->
             '-Xep:ChainedAssertionLosesContext:WARN',
             '-Xep:CharacterGetNumericValue:WARN',
             '-Xep:ClassCanBeStatic:WARN',
+            // '-Xep:ClassInitializationDeadlock:WARN', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:292]: This is probably useful

File: gradle/validation/error-prone.gradle [hunk 8]
```diff
@@ -296,6 +301,8 @@ allprojects { prj ->
             '-Xep:DateChecker:WARN',
             '-Xep:DateFormatConstant:WARN',
             // '-Xep:DefaultCharset:OFF', // we have forbiddenapis for that
+            //'-Xep:DeeplyNested:WARN', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:304]: This might be useful

File: gradle/validation/error-prone.gradle [hunk 9]
```diff
@@ -296,6 +301,8 @@ allprojects { prj ->
...
+            //'-Xep:DefaultLocale:WARN', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:305]: This is not needed. its handled by forbidden apis

File: gradle/validation/error-prone.gradle [hunk 10]
```diff
@@ -309,6 +316,7 @@ allprojects { prj ->
             '-Xep:EmptyBlockTag:WARN',
             // '-Xep:EmptyCatch:OFF', // todo check if useful or comment why not - might be handled by ECJ?
             // '-Xep:EmptySetMultibindingContributions:OFF', // we don't use this annotation
+            // '-Xep:EnumOrdinal:WARN', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:319]: This might be useful

File: gradle/validation/error-prone.gradle [hunk 11]
```diff
@@ -330,6 +338,7 @@ allprojects { prj ->
             // '-Xep:FragmentNotInstantiable:OFF', // we don't use android
             // '-Xep:FutureReturnValueIgnored:OFF', // todo there are problems that should be fixed
             '-Xep:GetClassOnEnum:WARN',
+            // '-Xep:GuiceNestedCombine:WARN', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:341]: We don't use guice

File: gradle/validation/error-prone.gradle [hunk 12]
```diff
@@ -383,6 +392,7 @@ allprojects { prj ->
             '-Xep:JodaPlusMinusLong:WARN',
             '-Xep:JodaTimeConverterManager:WARN',
             '-Xep:JodaWithDurationAddedLong:WARN',
+            // '-Xep:JUnitIncompatibleType:WARN', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:395]: This is probably useful

File: gradle/validation/error-prone.gradle [hunk 13]
```diff
@@ -403,10 +413,12 @@ allprojects { prj ->
             // '-Xep:MissingSummary:OFF', // style preference that we don't want to enforce
             // '-Xep:MixedMutabilityReturnType:OFF', // todo check if useful or comment why not
             '-Xep:MockNotUsedInProduction:WARN',
+            // '-Xep:MockitoDoSetup:WARN', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:416]: This might be useful

File: gradle/validation/error-prone.gradle [hunk 14]
```diff
@@ -403,10 +413,12 @@ allprojects { prj ->
...
             '-Xep:ModifiedButNotUsed:WARN',
             '-Xep:ModifyCollectionInEnhancedForLoop:WARN',
             '-Xep:ModifySourceCollectionInStream:WARN',
             '-Xep:MultimapKeys:WARN',
+            // '-Xep:MultipleNullnessAnnotations:WARN', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:421]: We don't use nullness annotations but might be useful

File: gradle/validation/error-prone.gradle [hunk 15]
```diff
@@ -428,7 +440,9 @@ allprojects { prj ->
             '-Xep:NullableOptional:WARN',
             // '-Xep:NullablePrimitive:OFF', // we don't use this annotation
             // '-Xep:NullablePrimitiveArray:OFF', // we don't use this annotation
+            // '-Xep:NullableTypeParameter:WARN', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:443]: Might be useful

File: gradle/validation/error-prone.gradle [hunk 16]
```diff
@@ -428,7 +440,9 @@ allprojects { prj ->
...
             // '-Xep:NullableVoid:OFF', // we don't use this annotation
+            // '-Xep:NullableWildcard:WARN', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:445]: Might be useful

File: gradle/validation/error-prone.gradle [hunk 17]
```diff
@@ -442,6 +456,7 @@ allprojects { prj ->
             '-Xep:Overrides:WARN',
             // '-Xep:OverridesGuiceInjectableMethod:OFF', // we don't use guice
             '-Xep:ParameterName:WARN',
+            // '-Xep:PatternMatchingInstanceof:WARN', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:459]: Not sure we need this since its subjective

File: gradle/validation/error-prone.gradle [hunk 18]
```diff
@@ -459,6 +474,7 @@ allprojects { prj ->
             // '-Xep:SameNameButDifferent:OFF', // todo check if useful or comment why not
             '-Xep:SelfAlwaysReturnsThis:WARN',
             // '-Xep:ShortCircuitBoolean:OFF', // todo check if useful or comment why not
+            // '-Xep:StatementSwitchToExpressionSwitch:WARN', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:477]: Not sure this is needed since its subjective

File: gradle/validation/error-prone.gradle [hunk 19]
```diff
@@ -469,9 +485,12 @@ allprojects { prj ->
             '-Xep:StringCharset:WARN',
             '-Xep:StringFormatWithLiteral:WARN',
             // '-Xep:StringSplitter:OFF', // todo check if useful or comment why not - might be able to use forbidden-apis for this?
+            // '-Xep:SunApi:WARN', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:488]: This is probably useful

File: gradle/validation/error-prone.gradle [hunk 20]
```diff
@@ -469,9 +485,12 @@ allprojects { prj ->
...
+            // '-Xep:SuperCallToObjectMethod:WARN', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:489]: This is probably useful

File: gradle/validation/error-prone.gradle [hunk 21]
```diff
@@ -469,9 +485,12 @@ allprojects { prj ->
...
             '-Xep:SuperEqualsIsObjectEquals:WARN',
             // '-Xep:SwigMemoryLeak:OFF', // we don't use swig
             // '-Xep:SynchronizeOnNonFinalField:OFF', // todo check if useful or comment why not
+            // '-Xep:SystemConsoleNull:WARN', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:493]: This is probably useful

File: gradle/validation/error-prone.gradle [hunk 22]
```diff
@@ -493,6 +512,7 @@ allprojects { prj ->
             // '-Xep:UnicodeEscape:OFF', // can't enable since Lucene/Solr tests use unicode a bunch
             // '-Xep:UnnecessaryAssignment:OFF', // we don't use these annotations
             '-Xep:UnnecessaryAsync:WARN',
+            // '-Xep:UnnecessaryBreakInSwitch:WARN', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:515]: Not sure this is useful or just subjective

File: gradle/validation/error-prone.gradle [hunk 23]
```diff
@@ -513,6 +533,7 @@ allprojects { prj ->
             // '-Xep:UseBinds:OFF', // we don't use this annotation
             // '-Xep:UseCorrectAssertInTests:OFF', // we inherit from LuceneTestCase which extends Assert
             '-Xep:VariableNameSameAsType:WARN',
+            // '-Xep:VoidUsed:WARN', // todo check if useful or comment why not
```
Comment (risdenk) [inline gradle/validation/error-prone.gradle:536]: Might be useful

**Review by risdenk — COMMENTED** [review]
Added notes going through the new rules. Most can be ignored.

**Comment by malliaridis** [comment]
Thanks for the notes @risdenk. I will soon create another PR for applying the rules and see what can be backported as well. We can then address the existing rules as well. Is it fine to do it under the context of [SOLR-16427](https://issues.apache.org/jira/browse/SOLR-16427)?

**Pull request merged**